partition.go

```diff
@@ -146,12 +147,23 @@ func packageNameFromPath(path string) (string, error) {
 	}
 	for _, gopath := range getGoPathList() {
 		rel, err := filepath.Rel(filepath.Join(gopath, "src"), path)
-		if err != nil {
+		if err != nil || rel[0] == '.' {
 			continue
 		}
 		return rel, nil
 	}
-	return "", fmt.Errorf("%s not in GOPATH", path)
+
+	wd, err := os.Getwd()
```

Comment:
I don't think the rest of this is necessary. Paths outside of gopath shouldn't be supported.

Reply:
But as far as I know, all the linters work well in the outside of GOPATH case, and they all correctly indicate the program's error, or exit quietly if not found any errors.

Please have a look(called with my modifications):
```
$ pwd
/Users/flw/study/go/foo
$ go env GOPATH
/Users/flw/.go
$ gometalinter --enable-all good-code/hello.go && echo 'OK' || echo 'BAD'
OK
$ gometalinter --enable-all bad-code/hello.go && echo 'OK' || echo 'BAD'
bad-code/hello.go:6:2:error: undeclared name: foo (gotypex)
bad-code/hello.go:6:2:warning: unused struct field undeclared name: foo (structcheck)
bad-code/hello.go:6:2:warning: undeclared name: foo (megacheck)
bad-code/hello.go:6:2:error: undefined: foo (vet)
bad-code/hello.go:6:2:warning: undefined: foo (vetshadow)
bad-code/hello.go:6:2:warning: error return value not checked (undeclared name: foo) (errcheck)
bad-code/hello.go:6:2:warning: parameter undeclared name: foo (unparam)
bad-code/hello.go:6:2:warning: unused variable or constant undeclared name: foo (varcheck)
bad-code/hello.go:6:2:error: undeclared name: foo (gotype)
bad-code/hello.go:6:2:warning: undeclared name: foo (interfacer)
BAD
$ cat good-code/hello.go
package main

import "fmt"

func main() {
        fmt.Println("vim-go")
}
$ cat bad-code/hello.go
package main

import "fmt"

func main() {
        foo = 1
        fmt.Println("vim-go")
}
$
```

Reply:
GOPATH or NOT GOPATH? This is a problem. 
Different people have different opinions on this. Let us suspend that disputes temporarily, take a look at what [godoc](https://golang.org/cmd/go/#hdr-Import_path_syntax) said:

> if you are compiling a Go program not in a work space, you can use a relative path in an import statement in that program to refer to nearby code also not in a work space. This makes it easy to experiment with small multipackage programs outside of the usual work spaces.....

In my understanding, golang supports `outside of GOPATH` explicitly.

A few weeks ago, rsc made [a proposal](https://research.swtch.com/vgo-intro) to explicitly deprecates GOPATH:

> This proposal keeps the best parts of go get, adds reproducible builds, adopts semantic versioning, eliminates vendoring, deprecates GOPATH in favor of a project-based workflow

And this proposal has been [accepted](https://research.swtch.com/vgo-accepted) at this moment. 

I hope above information will help everyone make correct judgments.

Reply:
I don't think your example is accurate because it doesn't try to import any code outside of the GOPATH. Sure a single package might work, but that is probably not the common case. Linters that operate on file paths will probably work, but anything that uses `x/tools/go/loader` will fail I believe.

As for your other arguments:

> take a look at what godoc

That quote says "you can use a relative path in an import statement ...  to experiment".  I think it's very unlikely anyone is running linters against experimental code. In practice they use absolute paths in non-experimental code (not relative paths), so that quote is not really relevant to this use case.

> golang supports outside of GOPATH explicitly.
This is a very useful feature. When I want to quickly experience a piece of code, I can vim foo.go, then go run foo.go, and get the answer.

Again, this is only if you're developing a `main` package. As soon as you have multiple packages this breaks.

> A few weeks ago, rsc made a proposal to explicitly deprecate GOPATH:

I've aware of the vgo proposal. However it's not implemented in an official release yet. When it is, `x/tools/go/loader` should work outside of a GOPATH, so linters will start to work outside of a go path, and then it should be fine to support it.